File: GUI/dynamic/parser.py
"""
Real-time stream parser for embedded tool calls.

The model emits tags like:
    <TOOL_CALL>{"action": "set_background", "color": "#1e1e2e"}</TOOL_CALL>

Feed tokens in one character at a time via feed(). The parser yields:
    ("text",   str)  — characters safe to display
    ("action", dict) — parsed JSON payload ready for the dispatcher

The lookahead buffer holds up to len(OPEN) chars so regular text is
only delayed by at most one tag-length, then flushed immediately.
"""
import json

OPEN = "<TOOL_CALL>"
CLOSE = "</TOOL_CALL>"


class ActionStreamParser:
    def __init__(self) -> None:
        self._buf = ""      # potential OPEN-tag prefix accumulator
        self._in_tag = False
        self._tag_buf = ""  # content inside an open tag

    def feed(self, chunk: str) -> list[tuple[str, object]]:
        events: list[tuple[str, object]] = []
        for c in chunk:
            if self._in_tag:
                self._tag_buf += c
                if self._tag_buf.endswith(CLOSE):
                    payload = self._tag_buf[: -len(CLOSE)]
                    try:
                        events.append(("action", json.loads(payload)))
                    except json.JSONDecodeError:
                        events.append(("text", OPEN + payload + CLOSE))
                    self._tag_buf = ""
                    self._in_tag = False
            else:
                self._buf += c
                if self._buf.endswith(OPEN):
                    pre = self._buf[: -len(OPEN)]
                    if pre:
                        events.append(("text", pre))
                    self._buf = ""
                    self._in_tag = True
                # Don't emit char-by-char — bulk-flush after the loop.

        # Emit everything that can't possibly be the start of OPEN.
        # Keep only OPEN.length-1 chars as lookahead.
        if not self._in_tag and len(self._buf) > len(OPEN) - 1:
            safe = len(self._buf) - (len(OPEN) - 1)
            events.append(("text", self._buf[:safe]))
            self._buf = self._buf[safe:]

        return events

    def flush(self) -> list[tuple[str, object]]:
        events: list[tuple[str, object]] = []
        if self._buf:
            events.append(("text", self._buf))
        if self._in_tag and self._tag_buf:
            events.append(("text", OPEN + self._tag_buf))
        self._buf = ""
        self._tag_buf = ""
        self._in_tag = False
        return events

File: GUI/frontend/src/dynamic/parser.ts
/**
 * Real-time stream parser for embedded tool calls.
 *
 * The model emits:
 *   <TOOL_CALL>{"action": "set_theme", "name": "ocean"}</TOOL_CALL>
 *
 * feed() processes tokens one chunk at a time and yields:
 *   { type: 'text',   value: string }
 *   { type: 'action', value: Record<string, unknown> }
 *
 * Tags are stripped from the text stream before reaching the display.
 */

const OPEN  = '<TOOL_CALL>';
const CLOSE = '</TOOL_CALL>';

export type StreamEvent =
  | { type: 'text';   value: string }
  | { type: 'action'; value: Record<string, unknown> };

export class ActionStreamParser {
  private buf    = '';   // lookahead for potential OPEN prefix
  private inTag  = false;
  private tagBuf = '';   // content inside an open tag

  feed(chunk: string): StreamEvent[] {
    const events: StreamEvent[] = [];
    for (const c of chunk) {
      if (this.inTag) {
        this.tagBuf += c;
        if (this.tagBuf.endsWith(CLOSE)) {
          const payload = this.tagBuf.slice(0, -CLOSE.length);
          try {
            events.push({ type: 'action', value: JSON.parse(payload) as Record<string, unknown> });
          } catch {
            events.push({ type: 'text', value: OPEN + payload + CLOSE });
          }
          this.tagBuf = '';
          this.inTag  = false;
        }
      } else {
        this.buf += c;
        if (this.buf.endsWith(OPEN)) {
          const pre = this.buf.slice(0, -OPEN.length);
          if (pre) events.push({ type: 'text', value: pre });
          this.buf   = '';
          this.inTag = true;
        }
        // Don't emit char-by-char — bulk-flush after the loop instead.
      }
    }

    // After consuming the whole chunk, emit everything that provably can't be
    // the start of an OPEN tag. We only need to keep OPEN.length-1 chars as
    // lookahead (a complete OPEN match would already have been caught above).
    if (!this.inTag && this.buf.length > OPEN.length - 1) {
      const safe = this.buf.length - (OPEN.length - 1);
      events.push({ type: 'text', value: this.buf.slice(0, safe) });
      this.buf = this.buf.slice(safe);
    }

    return events;
  }

  flush(): StreamEvent[] {
    const events: StreamEvent[] = [];
    if (this.buf)                        events.push({ type: 'text', value: this.buf });
    if (this.inTag && this.tagBuf)       events.push({ type: 'text', value: OPEN + this.tagBuf });
    this.buf    = '';
    this.tagBuf = '';
    this.inTag  = false;
    return events;
  }
}
